TypedSyntax/src/show.jl

```diff
@@ -113,12 +113,21 @@ function type_annotation_mode(node, @nospecialize(T); type_annotations::Bool, hi
     return type_annotate, pre, pre2, post
 end
 
+"""
+    type_string(type)
+
+Overload this function to implement custom annotations for types. For example,
+`type_string(::Type{Float32}) = "F32"` would result in `F32` being used to annotate
+any instance of a `Float32`.
+"""
+type_string(T) = string(T)
+
 function show_annotation(io, @nospecialize(T), post, node, position; iswarn::Bool)
     diagnostics = get(io, :diagnostics, nothing)
     inlay_hints = get(io, :inlay_hints, nothing)
 
     print(io, post)
-    T_str = string(T)
+    T_str = type_string(T)
```

Comment:
Do we need to make this recursive? I think it would be weird if we printed like this:
```julia
Float32: "F32"
Vector{Float32}: "Vector{Float32}"
```

Reply:
I think it might be hard to do that generally – I just assumed the user could set that up for all the types they'd be interested in. Otherwise it would get really complicated quickly. But maybe there's an easy way to apply it recursively?

Reply:
Maybe we could set it up for the basic ones though, like `Vector`, `Union`, `Core.Const`, etc.?

Reply:
Here's a basis for doing it systematically:

```julia
julia> T = Dict{String,Vector{Float32}}
Dict{String, Vector{Float32}}

julia> typeof(T)
DataType

julia> T.parameters
svec(String, Vector{Float32})

julia> T.parameters[1]
String

julia> T.parameters[2].parameters
svec(Float32, 1)
```

Reply:
Alternative thought: generate the string and `replace`?

Reply:
I see, let me try, thanks.

Re: strings, I've tried this in a different context but it ended up being a lot of trouble because regexp isn't Turing complete. For example there's no way to do nested parentheses matching with regexp. After that experience I've mostly stuck to converting to a string as late as possible.